hphp/util/asm-x64-legacy.h

```diff
@@ -0,0 +1,1389 @@
+  BYTE_LOAD_OP(name, instr##b)
+
+#define BYTE_STORE_OP(name, instr)                                    \
+  void name##b(Reg8 r, MemoryRef m)        { instrRM(instr, r, m); }  \
+  void name##b(Immed i, MemoryRef m)       { instrIM8(instr, i, m); } \
+
+#define STORE_OP(name, instr)                                           \
+  void name##w(Immed i, MemoryRef m) { instrIM16(instr, i, m); }        \
+  void name##l(Immed i, MemoryRef m) { instrIM32(instr, i, m); }        \
+  void name##w(Reg16 r, MemoryRef m) { instrRM(instr, r, m); }          \
+  void name##l(Reg32 r, MemoryRef m) { instrRM(instr, r, m); }          \
+  void name##q(Reg64 r, MemoryRef m) { instrRM(instr, r, m); }          \
+  BYTE_STORE_OP(name, instr ## b)
+
+#define BYTE_REG_OP(name, instr)                              \
+  void name##b(Reg8 r1, Reg8 r2) { instrRR(instr, r1, r2); }  \
+  void name##b(Immed i, Reg8 r)  { instrIR(instr, i, r); }    \
+
+#define REG_OP(name, instr)                                       \
+  void name##q(Reg64 r1, Reg64 r2)   { instrRR(instr, r1, r2); }  \
+  void name##l(Reg32 r1, Reg32 r2)   { instrRR(instr, r1, r2); }  \
+  void name##w(Reg16 r1, Reg16 r2)   { instrRR(instr, r1, r2); }  \
+  void name##l(Immed i, Reg32 r)     { instrIR(instr, i, r); }    \
+  void name##w(Immed i, Reg16 r)     { instrIR(instr, i, r); }    \
+  BYTE_REG_OP(name, instr##b)
+
+  /*
+   * For when we a have a memory operand and the operand size is
+   * 64-bits, only a 32-bit (sign-extended) immediate is supported.
+   */
+#define IMM64_STORE_OP(name, instr)             \
+  void name##q(Immed i, MemoryRef m) {          \
+    return instrIM(instr, i, m);                \
+  }
+
+  /*
+   * For instructions other than movq, even when the operand size is
+   * 64 bits only a 32-bit (sign-extended) immediate is supported.
+   */
+#define IMM64R_OP(name, instr)                  \
+  void name##q(Immed imm, Reg64 r) {            \
+    always_assert(imm.fits(sz::dword));         \
+    return instrIR(instr, imm, r);              \
+  }
+
+#define FULL_OP(name, instr)                    \
+  LOAD_OP(name, instr)                          \
+  STORE_OP(name, instr)                         \
+  REG_OP(name, instr)                           \
+  IMM64_STORE_OP(name, instr)                   \
+  IMM64R_OP(name, instr)
+
+  // We rename x64's mov to store and load for improved code
+  // readability.
+  LOAD_OP        (load,  instr_mov)
+  STORE_OP       (store, instr_mov)
+  IMM64_STORE_OP (store, instr_mov)
+  REG_OP         (mov,   instr_mov)
+
+  FULL_OP(add, instr_add)
+  FULL_OP(xor, instr_xor)
+  FULL_OP(sub, instr_sub)
+  FULL_OP(and, instr_and)
+  FULL_OP(or,  instr_or)
+  FULL_OP(test,instr_test)
+  FULL_OP(cmp, instr_cmp)
+  FULL_OP(sbb, instr_sbb)
+
+#undef IMM64_OP
+#undef IMM64R_OP
+#undef FULL_OP
+#undef REG_OP
+#undef STORE_OP
+#undef LOAD_OP
+#undef BYTE_LOAD_OP
+#undef BYTE_STORE_OP
+#undef BYTE_REG_OP
+#undef IMM64_STORE_OP
+#undef IMM64R_OP
```

Comment:
Drop this -- same as line 295 above.

Reply:
https://github.com/facebook/hhvm/pull/8164/commits/ac4a93085f536b08c1f366528bb51a4599d8b290